File: src/watchdog.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const controller = require('./controller');
const { stateDir, watchdogFile, logFile, debugEnabled } = require('./paths');

const TICK_MS = 30000;

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

function log(message) {
  if (!debugEnabled()) return;
  try {
    const time = new Date().toISOString().slice(11, 19);
    fs.appendFileSync(logFile(), `${time} pid=${process.pid} watchdog: ${message}\n`);
  } catch {
  }
}

function idleTimeoutMs() {
  const seconds = Number(process.env.VIBECODE_IDLE_TIMEOUT);
  return (seconds > 0 ? seconds : 600) * 1000;
}

function abandoned(idleMs, limitMs) {
  return idleMs >= limitMs;
}

function beat() {
  try {
    fs.mkdirSync(stateDir(), { recursive: true });
    fs.writeFileSync(watchdogFile(), String(process.pid));
  } catch {
  }
}

function cleanup() {
  try {
    fs.unlinkSync(watchdogFile());
  } catch {
  }
}

async function run() {
  log(`started (timeout ${idleTimeoutMs() / 1000}s)`);
  for (;;) {
    beat();
    await sleep(TICK_MS);
    const status = await controller.status();
    if (status === '') {
      log('player gone, exiting');
      return cleanup();
    }
    const idleMs = Date.now() - controller.lastActivityMs();
    if (abandoned(idleMs, idleTimeoutMs())) {
      log(`abandoned ${Math.round(idleMs / 1000)}s, stopping`);
      await controller.hardPause();
      return cleanup();
    }
  }
}

module.exports = { abandoned, TICK_MS };

if (require.main === module) {
  run().catch(() => cleanup());
}

File: src/watchdog.rs
use crate::controller;
use crate::log::log;
use crate::paths;
use std::env;
use std::fs;
use std::thread::sleep;
use std::time::Duration;

const TICK_MS: u64 = 30000;

fn idle_timeout_ms() -> i64 {
    let secs = env::var("VIBECODE_IDLE_TIMEOUT")
        .ok()
        .and_then(|s| s.parse::<i64>().ok())
        .unwrap_or(0);
    (if secs > 0 { secs } else { 600 }) * 1000
}

fn now_ms() -> i64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as i64
}

pub fn run() {
    log(&format!("started (timeout {}s)", idle_timeout_ms() / 1000));
    loop {
        let _ = fs::create_dir_all(paths::state_dir());
        let _ = fs::write(paths::watchdog_file(), std::process::id().to_string());
        sleep(Duration::from_millis(TICK_MS));
        if controller::status().is_empty() {
            log("player gone, exiting");
            break;
        }
        let idle = now_ms() - controller::last_activity_ms();
        if idle >= idle_timeout_ms() {
            log(&format!("abandoned {}s, quitting mpv", idle / 1000));
            controller::quit_player();
            break;
        }
    }
    let _ = fs::remove_file(paths::watchdog_file());
}
